File: src/js/TicTacToeTrainer.js
/*
	This is a model that represents TicTacToe game status for training purpose
	There's only one AI used here - and user play tictactoe with it.
	And this model can have a list of AI's strategies but cannot run them on client-side for security reason.   
	For each move of AI, this model queries(by sending current board) to the server and retrieves 
		a list of all the AI's strategies(code, title, applicability and list of locations) in the order of priority,
	For each move of user, this model queries(by sending current board and user's move) to the server and gets
		a list of all the public strategies(code, title, whether AI knows or not) in random order. 
	When user confirms a new strategy is used for his move, 
		it sends a query (the rule's code) to update AI's rule set
	
	Each game done here is not recorded.  
		 	
oct13 2011, by Tak
*/


function TicTacToeTrainer() {		
	/* BASIC FUNCTIONS */
    this.createEmptyBoard = function (col, row) {
        var b = new Array();
        for (var i = 0; i < row; i++) {
            var r = new Array();
            for (var j=0;j<col;j++) {
            	r.push(0);
            }
            b.push(r);
        }
        return b;
    }
	this.flipTurn = function(player) {
		if (player==this.p1) return this.p2
		else return this.p1
	}
	this.cloneBoard = function(b) {
		var nb = this.createEmptyBoard(this.width,this.height);
		for(var i=0;i<this.width;i++) {
			for (var j=0;j<this.height;j++) {
				nb[i][j] = b[i][j];
			}
		}
		return nb;
	}
    this.restart = function() {
        this.firstTurn = this.flipTurn(this.firstTurn);
    	this.turn = this.firstTurn;
        this.board = this.createEmptyBoard(this.width, this.height);
        this.history = [];
        this.history.push({'board':this.cloneBoard(this.board), 'loc':undefined, 'turn':undefined});
        this.getStrategy(this.user,function(response){
        	this.strategy = response.data;
        });
        this.started=false;
    }
//  AS THE SERVER HAS ALL THE RULE OF THE USER, trainer doesn't update all the strategies at once,
//	this.updateRule = function() {
//		// 
//		
// 		var user_id = this.p1;
// 		var game_id = 'tictactoe';
// 		var strategy = JSON.stringify(this.getEnabledStrategy());
// 		$.get('updateRule',
// 			{ 	player: user_id, 
// 				game: game_id,
// 				strategy: strategy 
// 			}, function(response) {
// 				alert(response);
// 			}	
// 		);
// 	}
 	this.resumeFromHistoryMode = function(step) {
 		if (step>=this.history.length-1) {
 			alert('history out of bound');	
 		} else {
 			resumePoint = this.history[step];
 			this.board = resumePoint['board'];
 			this.turn = this.flipTurn(resumePoint['turn']); // turn of each history is the one who made the previous move. 
 			this.history = this.history.slice(0,step); // cut history after the resume point
 		}
 	}
 	// move function applies to both user and AI's move - it updates board data and history
    this.move = function(i, j, t) {
        if (this.turn != t) {
        	console.log("player turn doesn't match!" + t + "!=" + this.turn);	
        	return;
        } 
        if (this.board[i][j] != 0) {
        	console.log("cannot move on occupied cell [" + i+","+j+"]");	
        	return;    	
        }
       	if (this.started==false)   this.started = true;
        if (this.p1==this.turn) {
        	this.board[i][j]=this.p1;
        } else {
        	this.board[i][j]=this.p2;
        }
        this.history.push({'board':this.cloneBoard(this.board), 'loc':[i,j], 'turn':t});
        this.turn = this.flipTurn(this.turn);
        return this.board[i][j];
    }

    // FUNCTIONS FOR CHECKING CURRENT BOARD STATE //
    this.isFull = function() {
        for (var i = 0; i < 3; i++)
			for (var j =0; j<3;j++)
			    if (this.board[i][j] == 0)
			        return false;
        return true;
    }
	this.checkCol =function(x) {
        if (this.board[x][0] == this.board[x][1] && (this.board[x][0] == this.board[x][2]) && (this.board[x][0] != 0)) 
        	return this.board[x][0];
        else return false;
    }
    this.checkRow = function(y) {
        if (this.board[0][y] == this.board[1][y] && (this.board[0][y] == this.board[2][y]) && (this.board[0][y] != 0))
        	return this.board[0][y];
        else return false;
    }
    this.checkDiag1 = function()  {
        if (this.board[0][0] != 0 && this.board[0][0] == this.board[1][1] && this.board[0][0] == this.board[2][2]) 
        	return this.board[0][0];
        else return false; 
    }
    this.checkDiag2 = function() {
        if (this.board[2][0] != 0 && this.board[1][1] == this.board[2][0] && this.board[0][2] == this.board[2][0]) 
        	return this.board[2][0];
        else return false;
    }
    this.checkForWinner = function() {
        /* Check the top row */
        var result = false;
        for (var i=0; i < 3; i++) {
            result = this.checkRow(i);
            if (result != false)
                return result;
            result = this.checkCol(i);
            if (result != false)
                return result;
        }
        result = this.checkDiag1();
        if (result != false)
            return result;
        result = this.checkDiag2();
        if (result != false)
            return result;
        return false;
    }
//	this.analyzeMove = function(brd,player,userMoveLoc) {
//		// to check which strategies are matching with given brd,turn,loc
// 		// query the server.  The response is a list of strategies.
//	
//		
//		
//// USE SERVER-SIDE FUNCTION INSTEAD		
////		matchingStrategy = [];
////		for (key in this.strategySet) {
////			st = this.strategySet[key];
////			// console.log(st);
////			//console("ana" + brd,player);
////			result = this[st.code](brd,player);
////			if (true /*result['success']==true*/) {
////				for(i in result.loc) {
////					l = result.loc[i];
////					if(userMoveLoc[0]==l[0] && userMoveLoc[1]==l[1]) {
////						matchingStrategy.push({'name':st.name,'code':st.code});
////					}
////					
////				}				
////			}
////		}
//		// console.log(matchingStrategy);
//		return matchingStrategy;
//	}
	
    ////////////////////////////////////////////
    // FUNCTIONS THAT COMMUNICATE WITH SERVER //
    ////////////////////////////////////////////
    
    this.getStrategy = function(userName,callBack) {
    	$.ajax({
    		type : "GET",
    		url: "/ajaxTrainer",
    		async: false,	// browser will hold until it gets the response from server
    		cache: false,
    		data: 	{	action: 'getStrategy',
    					player: userName,
    					game: this.gameTitle
    				},
    		success: function(response) {
    			callBack(response);
    		}
    	});
    }
    this.getPublicStrategyDict = function(callBack) {	// list of all possible public rules
    	$.ajax({
    		type : "GET",
    		url: '/ajaxTrainer',
    		async: true,
    		cache: false,
    		data: {  	action: 'getPublicStrategyDict',
    					game: this.gameTitle 
    				},
    		success: function(response) {
    			callBack(response);
    		}
    	});
    }
    // ask server to run all the strategies the AI knows one-by-one and retrieve the result
    this.findBestStrategy = function(board,turn,callBack) {
    	$.ajax({
    		type : "GET",
    		url: "/ajaxTrainer",
    		async: true,
    		cache: false,
    		data: { action: 'findBestStrategy',
    				user:this.user,
    				player1: this.p1,
    				player2: this.p2,
    				game : this.gameTitle,
    				board: JSON.stringify(board),
    				turn: turn
    				},
    		success: function(response) {
    			callBack(response);
    		}
    	})
    }
    
    // ask server to infer which strategy (among all the public rules) might be the one user has used.
	this.findMatchingStrategy = function(brd,turn,userMoveLoc,callBack)  {  
		// matchingRules : {name:text, code:text, locList:array of [x,y]}
		$.ajax({
			type : "GET",
			url: "/ajaxTrainer",
			async: false,
			data: 	{ 	action: 'findMatchingStrategy',
						user: this.user,
						player1: this.p1,
						player2: this.p2,
						turn: turn, 
 	 					game: this.gameTitle,
 	 					board: JSON.stringify(brd),
 	 					loc: JSON.stringify(userMoveLoc)
 	 				},
 	 		success: function(response) {
 	 				callBack(response);	
 	 		}
		});
	}  
	
	// it just tells server to add a rule(by code) to user's AI's strategy
    this.enableStrategy = function(userName,key) {
    	$.ajax({
    		type : "GET",
    		url: "/ajaxTrainer",
    		async: true,
    		cache: false,
    		data: 	{ 	action: 'enableStrategy',
    					player: userName,
    					game: this.gameTitle,
    					strategyToEnable : key
    				},
    		success: function(response) {
//    					alert(response);
    					if (response=='True') {
    						$(cons.target).find("#mR_"+key).append("<span style='font-size:12px; color:#955;'> I learned this new rule!</span>");
    					} else if(response=='False') {
    						$(cons.target).find("#mR_"+key).append("<span style='font-size:12px; color:#595;'> Thanks! But I knew it already.</span>");
    					}
    				}
    	
    	});
    }
    
    
    this.changeOrder = function(nameList) {
    	$.ajax({
    		type : "GET",
    		url: "/ajaxTrainer",
    		async: true,
    		data: 	{ 	action: 'changeOrder',
    					player: this.user,
    					game: this.gameTitle,
    					newStrategy : JSON.stringify(nameList)
    				},
    	});
    }
    // callback functions for ajax calls
    this.setStrategy = function(data) {
    	this.strategy = JSON.parse(data);
    	this.strategyKeyList = [];
    	for (i in this.strategy)
    		this.strategyKeyList.push(this.strategy[i].key);

    }
    this.setPublicStrategyDict = function(data) {
    	this.publicStrategyDict = JSON.parse(data);
    }
    //
	this.makeNewStrategy = function(board, turn, ruleBoard, name, desc,
									translationInvariant, flipping, rowPermutation,
									columnPermutation,rotation){
		$.ajax({
			type : "GET",
			url : "/ajaxTrainer",
			async : true,
			data : 	{	action : 'makeNewStrategy',
						user: this.user,
						player1: this.p1,
						player2: this.p2,
						turn: turn, 
 	 					game: this.gameTitle,
 	 					board: JSON.stringify(board),
						ruleBoard : JSON.stringify(ruleBoard),
						title : name,
						desc : desc,
						translationInvariant : translationInvariant,
						flipping : flipping,
						rowPermutation : rowPermutation,
						columnPermutation : columnPermutation,
						rotation : rotation				
					}
		});
	}
	this.init = function(p1,p2) {
		this.user = p1;
		this.firstTurn = p1;
		this.gameTitle = 'tictactoe';
		this.p1 = p1;
		this.p2 = p2;
		this.width=3; 
		this.height=3;  
		var started; 	// true when playing game is on
		
		this.publicStrategyDict = {};  // list of code -> other information of all the public codes for the game
		this.strategy = [];  // list of rules that this user's AI knows
		this.strategyKeyList = [];  // list of keys only
		this.history = [];	// pop in/out current board state for undo/redo moves
	
		this.turn = this.p1;
		this.started = true;    
		this.board = this.createEmptyBoard(this.width, this.height);	
		this.history.push({'board':this.cloneBoard(this.board), 'loc':undefined, 'turn':undefined});
        this.getStrategy(this.user, $.proxy(this.setStrategy,this)); // $.proxy(function,scope) : will force the function to run within the scope 
        this.getPublicStrategyDict($.proxy(this.setPublicStrategyDict,this));
	}


	
} // end of TicTacToeTrainer

File: src/TicTacToeTrainer.py
'''
    We use this class for TicTacToe trainer.  Trainer needs it when AI decides next move, 
    inferring user's move using basic heuristics.   Board status and winning check will be done on client js.  
    
@author: reflect9
'''
from google.appengine.ext import db
import sys, copy
from random import choice
from django.utils import simplejson
from datastore import *

class Type:
    EMPTY = 0
    P1 = 1
    P2 = 2
    SELECTED = 3
    IGNORE = 4

class TicTacToeTrainer:
    def __init__(self,user=None,player1=None,player2=None,game='tictactoe',board=None,turn=None):
        self.user = user;
        self.p1 = player1
        self.p2 = player2
        self.game = game
        self.board = board
        self.turn = turn
        self.strategy = self.loadStrategy(self.user,self.game)
#        print >>sys.stderr, "trainer board created : " 
#        print >>sys.stderr, self.board

    ''' BASIC FUNCTIONS '''
    def loadStrategy(self,player,game):
#        playerKey = db.GqlQuery("SELECT * FROM User WHERE uid=:1",player).get().key()
#        gameKey = db.GqlQuery("SELECT * FROM Game WHERE title=:1",game).get().key()
        codeList = getUserStrategy(player,game)
        print >>sys.stderr,"codeList in loadStrategy : "+ json.dumps(codeList)
#        st = simplejson.loads(strategy.data)
        return codeList
    def flip(self,player):
        if player=='p1':
            return 'p2'
        elif player=='p2':
            return 'p1'
        elif player==self.p1:
            return self.p2
        else:
            return self.p1
    def isFull(self):
        for row in self.board:  
            for cell in row:
                if cell==0: return False
        return True
    def checkCol(self,x):
        if self.board[x][0]==self.board[x][1] and self.board[x][1]==self.board[x][2] and self.board[x][0]!=0:
            return self.board[x][0]
        else: return False
    def checkRow(self,y):
        if self.board[0][y]==self.board[1][y] and self.board[1][y]==self.board[2][y] and self.board[0][y]!=0:
            return self.board[0][y]
        else: return False
    def checkDiag1(self):
        if self.board[0][0]!=0 and self.board[0][0]==self.board[1][1] and self.board[1][1]==self.board[2][2]:
            return self.board[0][0]
        else: return False
    def checkDiag2(self):
        if self.board[2][0]!=0 and self.board[2][0]==self.board[1][1] and self.board[0][2]==self.board[1][1]:
            return self.board[2][0]
        else: return False      
    def checkWinner(self):
        for i in range(0,3):
            if self.checkRow(i) != False: 
                return self.checkRow(i)
            if self.checkCol(i) != False: 
                return self.checkCol(i)
        if self.checkDiag1()!=False: 
            return self.checkDiag1()
        if self.checkDiag2()!=False:
            return self.checkDiag2() 
        return False
#    def makeMove(self,x,y,turn):
#        if turn!=self.turn:
#            print >>sys.stderr,"turn doesn't match!"
#            return
#        if self.board[x][y] != 0:
#            print >>sys.stderr,"cannot move on occupied cell ["+str(x)+","+str(y)+"]"
#            return
#        self.board[x][y]=self.turn
#        return self.board
    def findBestStrategy(self):
        if self.isFull():       return {'message':"Tie Game",'locList':None}
        resultList = []
#        print >>sys.stderr, "self.strategy : "+ json.dumps(self.strategy)
        for st in self.strategy:
            # locals() provide a dictionary of all elements in local scope
            # locals()[functionName] gives a handler to the function
            # thus, below we execute local function whose name is st['code']
#            print st
            p1p2 = 'p1' if self.turn==self.p1 else 'p2'
            if (hasattr(self, st)):
                strategyMethodToCall =  getattr(self, st)
                result = strategyMethodToCall(self.board,p1p2)
            else:
                strategy = [s for s in getPublicStrategy(self.game) if s['name'] == st]
                result = self.evaluateCreatedStrategy(eval(strategy[0]['boardList']),self.board,p1p2)
                
            resultList.append({'st':st, 'result':result })
        return resultList
#        return {'message':"no matching strategy found",'locList':None}
    def findMatchingStrategy(self,userLoc):
        matchingStrategy = [];
        allPublicStrategy = getPublicStrategy(self.game)
        print >>sys.stderr, allPublicStrategy
        print >>sys.stderr, userLoc
        for st in allPublicStrategy:
            print >>sys.stderr, st
            if st['name'] in self.strategy: st['enabled']=True # PREVIOUSLY st['code']
            else:  st['enabled']=False
#            print >>sys.stderr, self.board
            p1p2 = 'p1' if self.turn==self.p1 else 'p2'
            print >>sys.stderr, "board : " + json.dumps(self.board) + p1p2
            if (st['code'] != ''):
                strategyMethodToCall = getattr(self, st['code'])
                result = strategyMethodToCall(self.board,p1p2)
            else:
                result = self.evaluateCreatedStrategy(eval(st['boardList']), self.board, p1p2)
            if result['success'] and result['loc']!=None:
                for loc in result['loc']:
                    if int(userLoc[0])==loc[0] and int(userLoc[1])==loc[1]:
                        matchingStrategy.append(st)
                    else:
                        print >>sys.stderr, str(userLoc) + "__" + str(loc)
        return matchingStrategy
    
    
    ''' STRATEGIES'''
    def takeWin(self,board,player):
        combos = [
                  [[0,0],[1,1],[2,2]],
                  [[0,2],[1,1],[2,0]],
                  
                  [[0,0],[0,1],[0,2]],
                  [[1,0],[1,1],[1,2]],
                  [[2,0],[2,1],[2,2]],
                  
                  [[0,0],[1,0],[2,0]],
                  [[0,1],[1,1],[2,1]],
                  [[0,2],[1,2],[2,2]]                  
                  ]
#        print >>sys.stderr, board[1]
        for case in combos:
            countEmptyCell = 0
            countPlayerCell = 0
            for pos in case:
#                print >>sys.stderr, pos
                if board[pos[0]][pos[1]]==player:
                    countPlayerCell = countPlayerCell+1
                if board[pos[0]][pos[1]]==0:
                    emptyCell = pos
                    countEmptyCell = countEmptyCell+1
            if countEmptyCell==1 and countPlayerCell==2:
                return {'success':True, 'loc':[emptyCell]}
        return {'success':False,'loc':None}
    def takeBlockWin(self,board,player):
        opponent = self.flip(player)
        return self.takeWin(board,opponent)
    def takeCenter(self,board,player):
        if board[1][1]==0:
            return {'success':True, 'loc':[[1,1]]}
        return {'success':False}
    def takeAnyCorner(self,board,player):
        combos = [[0,0],[2,2],[2,0],[0,2]]
        possibleMoves = []
        print >>sys.stderr, board
        for case in combos:
            if board[case[0]][case[1]]==0:
                possibleMoves.append(case)
        if len(possibleMoves)>0:
            return {'success':True, 'loc':possibleMoves}
        else:
            return {'success':False, 'loc':None}
    def takeAnySide(self,board,player):
        combos = [[0,1],[1,0],[1,2],[2,1]]
        possibleMoves = []
        for case in combos:
            if board[case[0]][case[1]]==0:
                possibleMoves.append(case)
        if len(possibleMoves)>0:
            return {'success':True, 'loc':possibleMoves}
        else:
            return {'success':False, 'loc':None}
    def takeRandom(self,board,player):
        possibleMoves = []
        for x in range(0,3):
            for y in range(0,3):
                if board[x][y]==0:  possibleMoves.append([x,y])
        if len(possibleMoves)>0:
            return {'success':True, 'loc':possibleMoves}
        else:
            return {'success':False, 'loc':None}
        
    def takeOppositeCorner(self,board,player):
        combos = [[[0,0],[2,2]],[[0,2],[2,0]]]
        for corners in combos:
            c1 = corners[0]
            c2 = corners[1]
            if board[c1[0]][c1[1]]!=0 and board[c2[0]][c2[1]]==0:
                return {'success':True, 'loc':[c2]}
            if board[c2[0]][c1[1]]!=0 and board[c1[0]][c2[1]]==0:
                return {'success':True, 'loc':[c1]}      
        return {'success':False}

    def evaluateCreatedStrategy(self, ruleBoardList, board, player):
        result = {'success':True,'loc':[]}
        pushFlag = False
        for ruleBoard in ruleBoardList:
            width = len(ruleBoard)
            height = len(ruleBoard[0])
            for i in range(len(board)-width+1):
                for j in range(len(board[0])-height+1):
                    for k in range(i,i+width):
                        for l in range(j,j+height):
                            tile = ruleBoard[k-i][l-j]
                            if (tile == Type.P1 and board[k][l] != player):
                                result['success'] = False
                            elif (tile == Type.P2 and board[k][l] != self.flip(player)):
                                result['success'] = False
                            elif (tile == Type.EMPTY and board[k][l] != Type.EMPTY):
                                result['success'] = False
                            elif (tile == Type.SELECTED and board[k][l] == Type.EMPTY):
                                result['loc'].append([k,l])
                                pushFlag = True
                    if pushFlag and not result['success']:
                        result['loc'].pop()
                    pushFlag = False
                    result['success'] = True
        if len(result['loc']) == 0:
            result['success'] = False
        return result
        
    def makeNewStrategy(self, ruleBoard, name, desc, translationInvariant,
                        flipping, rowPermutation, columnPermutation, rotation):
        def addRule(container, toBeAdded):
            if toBeAdded not in container:
                container.append(toBeAdded)
        def flip(ruleBoard, horizontally, vertically):
            flipped = []
            for i in range(len(ruleBoard)):
                flipped.append([0]*len(ruleBoard[0]))
            for i in range(len(ruleBoard)):
                for j in range(len(ruleBoard[0])):
                    xIndex = i
                    yIndex = j
                    if horizontally:
                        xIndex = len(ruleBoard)-i-1
                    if vertically:
                        yIndex = len(ruleBoard[0])-j-1
                    flipped[xIndex][yIndex] = ruleBoard[i][j]
            return flipped
        def rotate(ruleBoard):
            # rotated will start with transposed dimensions
            rotated = []
            for i in range(len(ruleBoard[0])):
                rotated.append([0]*len(ruleBoard))
            # now define the values appropriately
            for i in range(len(ruleBoard)):
                for j in range(len(ruleBoard[0])):
                    xIndex = j
                    yIndex = len(ruleBoard)-i-1
                    rotated[xIndex][yIndex] = ruleBoard[i][j]
            return rotated
        def colPermute(ruleBoard,offset):
            permuted = []
            for i in range(len(ruleBoard)):
                permuted.append([0]*len(ruleBoard[0]))
            print permuted
            print ruleBoard
            for i in range(len(ruleBoard)):
                for j in range(len(ruleBoard[0])):
                    xIndex = (i + offset) % len(ruleBoard)
                    permuted[xIndex][j] = ruleBoard[i][j]
                    print (i,",",j,": ",permuted)
            return permuted
        def rowPermute(ruleBoard,offset):
            permuted = []
            for i in range(len(ruleBoard)):
                permuted.append([0]*len(ruleBoard[0]))
            for i in range(len(ruleBoard)):
                for j in range(len(ruleBoard[0])):
                    yIndex = (i + offset) % len(ruleBoard[0])
                    permuted[i][yIndex] = ruleBoard[i][j]
            return permuted
        def minimize(ruleBoard):
            hMin = len(ruleBoard)
            hMax = 0
            vMin = len(ruleBoard[0])
            vMax = 0
            for i in range(len(ruleBoard)):
                for j in range(len(ruleBoard[0])):
                    if ruleBoard[i][j] != Type.IGNORE:
                        hMin = min(hMin,i)
                        hMax = max(hMax,i)
                        vMin = min(vMin,j)
                        vMax = max(vMax,j)
            width = hMax-hMin+1
            height = vMax-vMin+1

            newBoard = []
            for i in range(hMin,hMin+width):
                newBoard.append([])
                for j in range(vMin,vMin+height):
                    newBoard[i].append(ruleBoard[i][j])
            return newBoard

        # Here is the start of the actual processing function
        
        if translationInvariant:
            ruleBoard = minimize(ruleBoard)

        boardList = []
        boardList.append(ruleBoard)

        if flipping:
            for i in range(len(boardList)):
                print boardList
                addRule(boardList,flip(boardList[i],True,False))
                print boardList
                addRule(boardList,flip(boardList[i],False,True))
                print boardList
                addRule(boardList,flip(boardList[i],True,True))
                print boardList

        if rowPermutation:
            for i in range(len(boardList)):
                for j in range(len(boardList[i][0])):
                    addRule(boardList,rowPermute(boardList[i],j))

        if columnPermutation:
            for i in range(len(boardList)):
                for j in range(len(boardList[i])):
                    addRule(boardList,colPermute(boardList[i],j))

        print boardList

        if rotation:
            for i in range(len(boardList)):
                rotated = rotate(boardList[i])
                addRule(boardList,rotated)
                addRule(boardList,flip(boardList[i],True,True))
                addRule(boardList,flip(rotated,True,True))

        print boardList.__repr__()
        Strategy(name=name,code='',key_name=name,description=desc,game='tictactoe',boardList=boardList.__repr__()).put()
